File: 1798-max-number-of-k-sum-pairs/1798-max-number-of-k-sum-pairs.java
class Solution {
    public int maxOperations(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        int res = 0;
        for (int i : nums) {
            int target = k - i;
            if (map.containsKey(target)) {
                map.put(target, map.get(target) - 1);
                res++;
                if (map.get(target) == 0) {
                    map.remove(target);
                }
            } else {
                map.put(i, map.getOrDefault(i, 0) + 1);
            }
        }
        return res;
    }
}

File: 1798-max-number-of-k-sum-pairs/1798-max-number-of-k-sum-pairs.py
class Solution:
    def maxOperations(self, nums: List[int], k: int) -> int:
        nums.sort()
        left, right, res = 0, len(nums)-1, 0
        while left < right:
            if nums[left] + nums[right] == k:
                res += 1
                left += 1
                right -= 1
            elif nums[left] + nums[right] > k:
                right -= 1
            else:
                left += 1
        return res